### PR TITLE
feat(analytics): bring back PoE mode

### PR DESCRIPTION
**Pull request by mariusandra**
## Problem

We could offload the decision on what PoE mode to use for users (with sane defaults)

## Changes

- Creates a system of overridden `modifiers` on the `Team` model. Anything on `team.modifiers` will override the team's default modifiers (visible under `team.default_modifiers`), which in turn can be overridden query time.
- Modifiers are the same things you see under `/debug`. They're intended to be user-controllable to an extent.
- Brings back the PoE toggle, but as a radio button, directly controlling the HogQL modifier

image

Update: Cut out the first and the last option:

image

## Does this work well for both Cloud and self-hosted?

Of course

## How did you test this code?

- Tested modifier overrides and saving in the UI
- Added a test to make sure the field has effect when creating the team's modifiers.

### CODE REVIEW

**Comment by tiina303** [comment]
I'm likely missing some context, what's the motivation for doing this?

Two questions:
1. why show users 4 options - I thought we're going with 100% will be event.person_id + overrides table join
2. why do we want to allow users to pick `event.person_properties`? 🦜  perhaps there's a better way longer term without confusing users (and us) with two different notions of what a person property is, see https://docs.google.com/document/d/1bxB-KIv4byKl-sVodow9E3dlo4QwOztMGVFhPLa2O94/edit if we didn't expose that to users, then a migration would be simpler.

**Comment by mariusandra** [comment]
My goals with this are:
- reduce support load
- decrease the support load
- increase time we can spend not doing support

I'm impartial to how we get there. I currently added all the PoE settings we have, but I'm happy to cut the list if you want to completely eradicate some, like the "pdi" one. I've still seen some teams using it in the wild (what prompted this PR in the first place), so included it here.

I do want to get some toggle out, as this will directly reduce support. Either by letting me quickly switch modes for users without messing with feature flags, or (this is the bet) users will be able to solve more of their own problems by choosing the mode.

---

So answers to questions above:

> why show users 4 options

What would you like me to show?

> why do we want to allow users to pick `event.person_properties`

Users have asked for this. There are different cases where you want to use the data as it was. From a purely practical perspective, with `person_properties` your past dashboards stop changing if you're relying on any datapoints like user countries, membership levels, etc.

This also makes queries many times faster, and sometimes just purely possible/practical.

I missed the boat, why are we trying to get rid of it? What's the harm? People want it.

**Review by timgl — DISMISSED** [review]

File: posthog/api/team.py [hunk 1]
```diff
@@ -165,6 +165,8 @@ class Meta:
             "groups_on_events_querying_enabled",
             "inject_web_apps",
             "extra_settings",
+            "modifiers",
```
[the remark this answers is not in the document]
Comment (mariusandra) [reply posthog/api/team.py:168]: Good idea, added

File: posthog/test/test_middleware.py [hunk 2]
```diff
@@ -123,7 +123,7 @@ class TestAutoProjectMiddleware(APIBaseTest):
     @classmethod
     def setUpTestData(cls):
         super().setUpTestData()
-        cls.base_app_num_queries = 42
+        cls.base_app_num_queries = 43
```
Comment (mariusandra) [inline posthog/test/test_middleware.py:126]: An extra PoE check when making the modifiers. Soon all PoE checks will be consolidated.

File: frontend/src/scenes/settings/project/PersonsOnEvents.tsx [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+        label: 'Deprecated old PoE setting. May incorrectly aggregate users. You probably want one of the other options.',
+    },
+    {
+        value: 'person_id_override_properties_on_events',
+        label: 'Use ingestion-time person properties from the events table (faster)',
+    },
+    {
+        value: 'person_id_override_properties_joined',
+        label: 'Use current person properties from the persons table (slower)',
+    },
```
Comment (tiina303) [inline frontend/src/scenes/settings/project/PersonsOnEvents.tsx:24]: Proposal for wording: https://posthog.slack.com/archives/C0368RPHLQH/p1714689440221919?thread_ts=1714679592.721539&cid=C0368RPHLQH

- Deprecated: Use person ids and properties from the time of the event. May show higher unique user counts due to not using latest person ids. You probably want one of the other options.
- Use person properties from the time of the event. Faster queries. If person property is updated, then query results on past data won't change.
- Use latest person properties. Slower queries. If person property is updated, then query results on past data will change to reflect it.

**Comment by mariusandra** [comment]
I've updated the text, and now we have this:

image

**Review by webjunkie — APPROVED** [review]

**Comment by bretthoerner** [comment]
This is what I see on a customer's page, not sure if it's intended but the "no radio selected, so if I pick one can I go back?" was a little confusing to me:

image

**Comment by mariusandra** [comment]
1. The third option should always be there now.
2. That "no radio" version shows up when the "disabled" PoE option is selected. I was confidently told that that's impossible, so I didn't account for that 😅

**Comment by bretthoerner** [comment]
> 1. The third option should always be there now.

I may be misreading what you mean, but I checked that project (33405) again and only see the 2?

**Comment by mariusandra** [comment]
I just merged in this: https://github.com/PostHog/posthog/pull/22071 -- which made the third option always visible. Now just the 4th one ("disabled" aka old joins via person_distinct_ids) is not there as something you can select.

**Comment by tiina303** [comment]
> That "no radio" version shows up when the "disabled" PoE option is selected. I was confidently told that that's impossible, so I didn't account for that 😅

That's because of this: https://posthog.slack.com/archives/C0368RPHLQH/p1714474490473039?thread_ts=1713927252.575329&cid=C0368RPHLQH

**Comment by bretthoerner** [comment]
> I saw it live myself, so not sure what's causing the difference... 🤔

I see it on our own projects too, just not customers. (i.e. US 33405) 🤷‍♂️

**Comment by mariusandra** [comment]
So we've come full circle to the original PR? 😆

**Pull request merged**